guides/ci/travis.md

```diff
@@ -0,0 +1,364 @@
+
+```yaml
+version: '3'
+
+services:
+  db:
+    image: mysql:8.0.17
+    command: --default-authentication-plugin=mysql_native_password
+    restart: always
+    environment:
+      MYSQL_ALLOW_EMPTY_PASSWORD: 'yes'
+    container_name: conways_db
+    ports:
+      - 3306:3306
+
+volumes:
+  db:
+```
+
+and start the container with:
+
+```shell-session
+$ docker-compose up
+```
+
+Using your favourite database client, we need to create a `database` named `test` with a `table` called `worlds` with two columns: `id` and `living_cells` (`int`).
+
+And set the environment variable:
+
+```shell-session
+$ export DATABASE_URL="mysql://root@localhost/test"
+```
+
+**Running the `specs` locally** should output 3 examples successfully. But, if we push the changes then Travis CI will report the following error: `Unknown database 'test' (Exception)`
+So, we need to configure Travis CI to use a MySQL service **and also setup the database**:
+
+>It’s really **important** to notice that the lines we are adding to `.travis.yml` will depend exclusively on the development workflow we are using!
+> And remember that this is only an example using MySQL.
+
+```yaml
+# .travis.yml
+language: crystal
+crystal:
+  - latest
+
+env:
+  - DATABASE_URL="mysql://root@localhost/test"
+
+services:
+  - mysql
+
+before_install:
+  - mysql -e 'CREATE DATABASE IF NOT EXISTS test;'
+  - mysql -e 'CREATE TABLE `test`.`worlds` (`id` serial,`living_cells` int NOT NULL DEFAULT 0, PRIMARY KEY (id));'
+
+script:
+  - crystal spec
+```
+
+Pushing these changes will trigger Travis CI and the build should be successful!!
+
+We [use a `setup.sql` script](https://andidittrich.de/2017/06/travisci-setup-mysql-tablesdata-before-running-tests.html) to create a more readable `.travis.yml` file:
+
+```yaml
+# .travis.yml
+language: crystal
+crystal:
+  - latest
+
+env:
+  - DATABASE_URL="mysql://root@localhost/test"
+
+services:
+  - mysql
+
+before_install:
+  - mysql -u root --password="" < test-data/setup.sql
+
+script:
+  - crystal spec
+```
+
+and then, the file `./test-data/setup.sql` will look like this:
+
+```sql
+-- setup.sql
+CREATE DATABASE IF NOT EXISTS test;
+CREATE TABLE `test`.`worlds` (`id` serial,`living_cells` int NOT NULL DEFAULT 0, PRIMARY KEY (id));
+```
+
+## Caching
+
+If we read Travis CI job log, we will find that every time the job runs, Travis CI needs to fetch the libraries needed to run the application:
+
+```log
+Fetching https://github.com/crystal-lang/crystal-mysql.git
+Fetching https://github.com/crystal-lang/crystal-db.git
+```
+
+This takes time and on the other hand these libraries might not change as often as our application, so it looks like we may cache them and save time.
+
+Travis CI [uses caching](https://docs.travis-ci.com/user/caching/) to improve some parts of the building path. For our example. here is the new configuration file **with cache enabled**:
```

Comment:
```suggestion
Travis CI [uses caching](https://docs.travis-ci.com/user/caching/) to improve some parts of the building path. For our example. Here is the new configuration file **with cache enabled**:
```

Reply:
Someday this could be integrated directly in a `setup_cache` method at https://github.com/travis-ci/travis-build/blob/master/lib/travis/build/script/crystal.rb . But for now, this section is needed.